### PR TITLE
xcdiff 0.10.0 (new formula)

### LINKED ISSUE #3: Add homebrew formula
**Is your feature request related to a problem? Please describe.**
To ease the process of installing xcdiff, would be great to add homebrew support such that it can be installed via:

```sh
brew install xcdiff
```

**Describe the solution you'd like**
Add a homebrew formula for xcdiff.

### PR DESCRIPTION
**Pull request by marciniwanicki**
Closes https://github.com/bloomberg/xcdiff/issues/3

xcdiff is an extensible tool that finds differences between two .xcodeproj project files. It can be thought of as git diff for .xcodeproj files, which can be used directly from the command line as well as a library supporting your own set of tools.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/xcdiff.rb
Comment (SMillerDev) [inline Formula/xcdiff.rb:18]: We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Comment (marciniwanicki) [reply Formula/xcdiff.rb:18]: Hi @SMillerDev , many thanks for pointing it out. I was about to ask about the test. Xcdiff tests xcodeprojects and sadly those files are very long and verbose.
Does the test have access to the sources? In such case we could just use one of the existing fixtures. Alternatively we can download the fixture (not sure if that's acceptable).

Comment (SMillerDev) [reply Formula/xcdiff.rb:18]: Downloading the fixture sounds good. You'll need a resource block and you'll need to name it something like `homebrew-testdata` so our automation knows it's for internal use only.

Comment (marciniwanicki) [reply Formula/xcdiff.rb:18]: Many thanks @SMillerDev, I will do it today.

**Comment by chenrui333** [comment]
ping @marciniwanicki

**Comment by marciniwanicki** [comment]
Sorry for the delays. 
Failed "CI / tests (ubuntu-22.04, ghcr.io/homebrew/ubuntu22.04:master, --user=linuxbrew", is Linux support a hard requirement? 
xcdiff is a dev tool for iOS/macOS developers (ie requires Xcode, macOS etc). I somehow don't think it would bring much value to support Linux.

**Comment by SMillerDev** [comment]
If linux isn't supported you need to add `depends_on :macos`

**Comment by marciniwanicki** [comment]
Many thanks @SMillerDev, I've pushed "depends_on :macos". Version xcdiff 0.9.0 didn't support Xcode: 14.1 and that's probably why "CI / tests (13-arm64-3802970930-1) (pull_request) " is also failing. I might release 0.10.0 supporting 14.1 and update the formula.

**Comment by marciniwanicki** [comment]
Slightly surprised (most likely missing some context), added `depends_on xcode: "13.3"`, but still getting Xcode 14.1

```
CPU: octa-core 64-bit arm_firestorm_icestorm
[417](https://github.com/Homebrew/homebrew-core/actions/runs/3803330424/jobs/6469701144#step:6:419)
  Clang: 14.0.0 build 1400
[418](https://github.com/Homebrew/homebrew-core/actions/runs/3803330424/jobs/6469701144#step:6:420)
  Git: 2.37.1 => /usr/bin/git
[419](https://github.com/Homebrew/homebrew-core/actions/runs/3803330424/jobs/6469701144#step:6:421)
  Curl: 7.79.1 => /usr/bin/curl
[420](https://github.com/Homebrew/homebrew-core/actions/runs/3803330424/jobs/6469701144#step:6:422)
  macOS: 12.6-arm64
[421](https://github.com/Homebrew/homebrew-core/actions/runs/3803330424/jobs/6469701144#step:6:423)
  CLT: 14.1.0.0.1.1665256668
[422](https://github.com/Homebrew/homebrew-core/actions/runs/3803330424/jobs/6469701144#step:6:424)
  Xcode: 14.1
[423](https://github.com/Homebrew/homebrew-core/actions/runs/3803330424/jobs/6469701144#step:6:425)
  Rosetta 2: false
  ```

**Comment by SMillerDev** [comment]
That's a minimum requirement. If it doesn't support newer XCodes already that would be a reason for me to hold off on merging this. The first version of Xcode 14.1 is nearly 3 months old.

**Comment by marciniwanicki** [comment]
OK, that's fair. I will cat 0.10.0. Thank you @SMillerDev a lot!

**Comment by marciniwanicki** [comment]
@SMillerDev extended the tests and fixed CI issues. Should be ready for review.

File: Formula/xcdiff.rb
Comment (SMillerDev) [inline Formula/xcdiff.rb:39]: Does this work?
```suggestion
    project = "Fixtures/ios_project_1/Project.xcodeproj"
    resource("homebrew-testdata").stage do
      assert_match "\n", shell_output("#{bin}/xcdiff -p1 #{project} -p2 #{project} -d")
      out = shell_output("#{bin}/xcdiff -p1 #{project} -p2 #{project} -g BUILD_PHASES -t Project -v")
      assert_match "✅ BUILD_PHASES > \"Project\" target\n", out
    end
  end
```
Otherwise maybe this?
```suggestion
    project = "Fixtures/ios_project_1/Project.xcodeproj"
    diff_args = "-p1 #{project} -p2 #{project}"
    resource("homebrew-testdata").stage do
      assert_match "\n", shell_output("#{bin}/xcdiff #{diff_args} -d")
      out = shell_output("#{bin}/xcdiff #{diff_args} -g BUILD_PHASES -t Project -v")
      assert_match "✅ BUILD_PHASES > \"Project\" target\n", out
    end
  end
```

File: Formula/xcdiff.rb
Comment (SMillerDev) [inline Formula/xcdiff.rb:29]: Can we assert something more than a linebreak here?

**Comment by marciniwanicki** [comment]
Thanks @SMillerDev for the suggestion. Extracting the parameters makes sense 👍

File: Formula/xcdiff.rb
Comment (marciniwanicki) [reply Formula/xcdiff.rb:29]: This is the case when there is no diff as such breakline is the only output. One possible improvement would be to assert also the exit code, but I wasn't sure if there is a way to capture it when using `shell_output`.

Comment (SMillerDev) [reply Formula/xcdiff.rb:29]: If it's only that, this is the better option
```suggestion
     # assert no difference between projects
      assert_equal "\n", shell_output("#{bin}/xcdiff #{diff_args} -d")
```

**Review by SMillerDev — APPROVED** [review]

**Comment by marciniwanicki** [comment]
@SMillerDev is there anything else we need to do before landing?

**Pull request closed without merging**
